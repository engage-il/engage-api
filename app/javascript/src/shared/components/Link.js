// @flow
import * as React from 'react'
import { Link as RouterLink } from 'react-router-dom'
import { cx, css } from 'react-emotion'
import type { Rule } from 'glamor'
import { colors } from '@/styles'

type Props = {
  to?: string,
  onClick?: () => void,
  styles?: Rule,
  className?: string,
  children?: React.Node
}

export function Link ({
  to: url,
  onClick,
  className,
  children,
  ...otherProps
}: Props) {
  if (!url && !onClick) {
    return null
  }

  const linkProps = {
    className: cx(link, className),
    onClick,
    children,
    ...otherProps
  }

  // use anchor tags for absolute urls, otherwise use a router link
  if (!url || /https?:\/\//.test(url)) {
    return <a {...linkProps} href={url} target='_blank' />
  } else {
    // $FlowFixMe: update react-router-dom flow libdefs
    return <RouterLink {...linkProps} to={url} />
  }
}

const link = css`
  color: ${colors.primary};
  transition: color 0.1s;

  &:hover {
    color: ${colors.primaryHighlight};
  }
`

export type { Props as LinkProps }
