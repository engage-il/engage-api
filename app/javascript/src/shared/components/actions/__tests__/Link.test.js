/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import { defaultsDeep } from 'lodash'
import { Link } from '../Link'

// subject
let subject

const defaults = {
  className: 'test-class',
  children: 'test-children'
}

function loadSubject (props = {}) {
  subject = shallow(<Link {...defaultsDeep(props, defaults)} />)
}

// specs
afterEach(() => {
  subject = null
})

describe('#render', () => {
  it('shows an anchor with an absolute url', () => {
    loadSubject({ to: 'http://www.test.com' })
    expect(subject).toMatchSnapshot()
  })

  it('shows a route link with a relative url', () => {
    loadSubject({ to: '/bills' })
    expect(subject).toMatchSnapshot()
  })

  it('hides if there is no url or click handler', () => {
    loadSubject()
    expect(subject.get(0)).toBe(null)
  })
})
