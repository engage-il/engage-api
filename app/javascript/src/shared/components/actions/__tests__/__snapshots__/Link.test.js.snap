// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render shows a route link with a relative url 1`] = `
<Link
  className="css-o62wdr-link test-class"
  replace={false}
  to="/bills"
>
  test-children
</Link>
`;

exports[`#render shows an anchor with an absolute url 1`] = `
<a
  className="css-o62wdr-link test-class"
  href="http://www.test.com"
  rel="noopener noreferrer"
  target="_blank"
>
  test-children
</a>
`;
